engine/schema/src/main/java/com/cloud/vm/dao/VmStatsDaoImpl.java

```diff
@@ -0,0 +1,122 @@
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+package com.cloud.vm.dao;
+
+import java.util.Date;
+import java.util.List;
+
+import javax.annotation.PostConstruct;
+
+import org.springframework.stereotype.Component;
+
+import com.cloud.utils.db.Filter;
+import com.cloud.utils.db.GenericDaoBase;
+import com.cloud.utils.db.SearchBuilder;
+import com.cloud.utils.db.SearchCriteria;
+import com.cloud.utils.db.SearchCriteria.Op;
+import com.cloud.vm.VmStatsVO;
+
+@Component
+public class VmStatsDaoImpl extends GenericDaoBase<VmStatsVO, Long> implements VmStatsDao {
+
+    protected SearchBuilder<VmStatsVO> vmIdSearch;
+    protected SearchBuilder<VmStatsVO> vmIdTimestampGreaterThanEqualSearch;
+    protected SearchBuilder<VmStatsVO> vmIdTimestampLessThanEqualSearch;
+    protected SearchBuilder<VmStatsVO> vmIdTimestampBetweenSearch;
+    protected SearchBuilder<VmStatsVO> timestampSearch;
+
+    @PostConstruct
+    protected void init() {
+        vmIdSearch = createSearchBuilder();
+        vmIdSearch.and("vmId", vmIdSearch.entity().getVmId(), Op.EQ);
+        vmIdSearch.done();
+
+        vmIdTimestampGreaterThanEqualSearch = createSearchBuilder();
+        vmIdTimestampGreaterThanEqualSearch.and("vmId", vmIdTimestampGreaterThanEqualSearch.entity().getVmId(), Op.EQ);
+        vmIdTimestampGreaterThanEqualSearch.and("timestamp", vmIdTimestampGreaterThanEqualSearch.entity().getTimestamp(), Op.GTEQ);
+        vmIdTimestampGreaterThanEqualSearch.done();
+
+        vmIdTimestampLessThanEqualSearch = createSearchBuilder();
+        vmIdTimestampLessThanEqualSearch.and("vmId", vmIdTimestampLessThanEqualSearch.entity().getVmId(), Op.EQ);
+        vmIdTimestampLessThanEqualSearch.and("timestamp", vmIdTimestampLessThanEqualSearch.entity().getTimestamp(), Op.LTEQ);
+        vmIdTimestampLessThanEqualSearch.done();
+
+        vmIdTimestampBetweenSearch = createSearchBuilder();
+        vmIdTimestampBetweenSearch.and("vmId", vmIdTimestampBetweenSearch.entity().getVmId(), Op.EQ);
+        vmIdTimestampBetweenSearch.and("timestamp", vmIdTimestampBetweenSearch.entity().getTimestamp(), Op.BETWEEN);
+        vmIdTimestampBetweenSearch.done();
+
+        timestampSearch = createSearchBuilder();
+        timestampSearch.and("timestamp", timestampSearch.entity().getTimestamp(), Op.LT);
+        timestampSearch.done();
+
+    }
+
+    @Override
+    public List<VmStatsVO> findByVmId(long vmId) {
+        SearchCriteria<VmStatsVO> sc = vmIdSearch.create();
```

Comment:
these two lines are duplicated. probably you can extract them to single function

Reply:
@ravening 
When I went to apply this suggestion I realized that the two lines mentioned are only repeated twice more. So I think it will end up not being worth it, as the ideal would be to add a new method + javadoc + unit tests for that. We may consider this in the future if there is a need to add more duplicates of these lines. At the moment, I think it's fine.
Thanks for the review.